[AMDGPU] Handle MachineOperandType global address in SIFoldOperands.

isakhilesh opened the pull request:
This patch handles the global operand type properly, fixing the
bug : Assertion `(isFI() || isCPI() || isTargetIndex() ||
isJTI()) && "Wrong MachineOperand accessor"` failed.

Fixes SWDEV-504645

isakhilesh commented:
Work In Progress: working on adding a test.

arsenm commented on llvm/lib/Target/AMDGPU/SIFoldOperands.cpp:
```diff
@@ -1161,8 +1161,10 @@ void SIFoldOperandsImpl::foldOperand(
-        else
+        else if(OpToFold.isFI())
           UseMI->getOperand(1).ChangeToFrameIndex(OpToFold.getIndex());
+        else if(OpToFold.isGlobal())
+          return;
```
Missing space after if, but this should probably be an else return, this will just break in the same way on any other operand type

arsenm commented on llvm/lib/Target/AMDGPU/SIFoldOperands.cpp:
```diff
@@ -1159,6 +1159,8 @@ void SIFoldOperandsImpl::foldOperand(
 
         UseMI->setDesc(TII->get(AMDGPU::S_MOV_B32));
 
+        if (OpToFold.isGlobal())
+          return;
```
This is still going to break on any non-global values

arsenm reviewed (changes requested):
Please do not post or update new reviews without a test.

isakhilesh committed:
[AMDGPU] Handle MachineOperandType global address in SIFoldOperands.

Fixes SWDEV-504645.

This patch handles the global operand type properly fixing the
bug : Assertion `(isFI() || isCPI() || isTargetIndex() ||
isJTI()) && "Wrong MachineOperand accessor"' failed.

arsenm commented on llvm/test/CodeGen/AMDGPU/swdev-504645.ll:
```diff
@@ -0,0 +1,47 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 5
+; RUN: llc -mtriple=amdgcn -mcpu=gfx942 -verify-machineinstrs < %s | FileCheck %s
```
```suggestion
; RUN: llc -mtriple=amdgcn -mcpu=gfx942 < %s | FileCheck %s
```

isakhilesh replied:
Done.

arsenm commented on llvm/test/CodeGen/AMDGPU/swdev-504645.ll:
```diff
@@ -0,0 +1,47 @@
+; CHECK-NEXT:    s_waitcnt vmcnt(0)
+; CHECK-NEXT:    s_setpc_b64 s[30:31]
+.entry:
+  %0 = call i32 @llvm.amdgcn.reloc.constant(metadata !0)
+  call void @test_buffer_load_sgpr_plus_imm_offset_noflags(i32 %0)
```
```suggestion
  %reloc = call i32 @llvm.amdgcn.reloc.constant(metadata !0)
  call void @test_buffer_load_sgpr_plus_imm_offset_noflags(i32 %reloc)
```

Use named values in tests

arsenm replied:
Function name is weird, this is just any call with an sgpr argument?

dhruvachak replied:
> Function name is weird, this is just any call with an sgpr argument?

Should be any call with an i32. The test case is from llvm-reduce applied to the original LL from the fuzzer JIRA mentioned in the description.

arsenm replied:
llvm-reduce does not change the names of functions, you can fix the name to show intent

isakhilesh replied:
Done, I have used a named value and updated the function name.

arsenm commented on llvm/test/CodeGen/AMDGPU/swdev-504645.ll:
```diff
@@ -0,0 +1,47 @@
+
+declare void @test_buffer_load_sgpr_plus_imm_offset_noflags(i32 inreg)
+
+; Function Attrs: nocallback nofree nosync nounwind speculatable willreturn memory(none)
```
```suggestion
```

isakhilesh replied:
Done

dhruvachak commented on llvm/lib/Target/AMDGPU/SIFoldOperands.cpp:
```diff
@@ -1161,6 +1161,10 @@ void SIFoldOperandsImpl::foldOperand(
 
         if (OpToFold.isImm())
           UseMI->getOperand(1).ChangeToImmediate(OpToFold.getImm());
+        else if (OpToFold.isGlobal())
+          UseMI->getOperand(1).ChangeToGA(OpToFold.getGlobal(),
+                                          OpToFold.getOffset(),
+                                          OpToFold.getTargetFlags());
         else
           UseMI->getOperand(1).ChangeToFrameIndex(OpToFold.getIndex());
```
> Maybe swap these? Handle isFI in the else if. We could also include explicit check for isGlobal, and give up on other operand types

Swapping isImm() and isGlobal() seems fine, if that's what you meant. 

There is a UseMI->setDesc() before this conditional block, so it was not clear whether it was too late to give up because a change may have been made already?

It seemed minimum changes compared to the existing code if the existing "else" was kept untouched and a new ifGlobal() handling was added since that was the asserting case to start with.

arsenm replied:
Yes, the setDesc would need to be sunk if the other case were handled

isakhilesh replied:
Hi @arsenm, I had initially considered handling isFI() in else if, but as @dhruvachak mentioned, UseMI->setDesc() has potentially made changes already, so I am keeping it in the else block since there might be modifications required with respect to setDesc() as well.

arsenm replied:
First, I meant do if (isImm) {} else if (isFI()) {} { else global

But you could also just move the setDesc to avoid the side effect

isakhilesh replied:
I can modify the patch to the first approach you mention here. Since we enter this block only in case `FoldingImmLike` is true and this variable is a bool for either immediate, frame index or global values. The if-else if-else conditions focus on only these 3 operands so there won't be any edge case for the else condition as we enter the block when it is one of the 3 operands.

But I'm not clear on how/where should I move the setDesc() to, since it is almost at the top of the block and the if-else if-else is immediately after it.

The structure is ->

if(FoldingImmLike){
     if(execMaybeModifiedBeforeUse)
                   return;
     UseMI->setDesc() // at the top of the block, in case the exec flag immediately above is false
      if-else if-else block
      UseMI->removeOperand(2)
      return;
}

Where should I place setDesc() to avoid a side effect?

dhruvachak replied:
> I can modify the patch to the first approach you mention here. Since we enter this block only in case `FoldingImmLike` is true and this variable is a bool for either immediate, frame index or global values. The if-else if-else conditions focus on only these 3 operands so there won't be any edge case for the else condition as we enter the block when it is one of the 3 operands.
> 
> But I'm not clear on how/where should I move the setDesc() to, since it is almost at the top of the block and the if-else if-else is immediately after it.
> 
> The structure is ->
> 
> if(FoldingImmLike){ if(execMaybeModifiedBeforeUse) return; UseMI->setDesc() // at the top of the block, in case the exec flag immediately above is false if-else if-else block UseMI->removeOperand(2) return; }
> 
> Where should I place setDesc() to avoid a side effect?

I think setDesc() has to be sunk only if you bail out. How about the following?

if (isImm()) {}
else if (isFI()) {}
else {
  assert(isGlobal());
  ...
}
Then you don't need to move setDesc().

arsenm replied:
@dhruvachak is spelling out the same as I stated

isakhilesh replied:
Got it, I have updated the patch based on the suggested structure, thanks a lot. Please let me know if I can improve it further.

dhruvachak commented:
Occurred to me that perhaps the test name should be changed since the swdev# won't make sense to everyone. What do you think @arsenm ?

isakhilesh committed:
Patch with updated test case, based on feedback.

isakhilesh commented:
> > Occurred to me that perhaps the test name should be changed since the swdev# won't make sense to everyone. What do you think @arsenm ?
> 
> I usually include that as a suffix, or in the test function name with some other description

Yes, updated the name of the test file.

arsenm commented on llvm/lib/Target/AMDGPU/SIFoldOperands.cpp:
```diff
@@ -1161,6 +1161,10 @@ void SIFoldOperandsImpl::foldOperand(
 
         if (OpToFold.isImm())
           UseMI->getOperand(1).ChangeToImmediate(OpToFold.getImm());
+        else if (OpToFold.isGlobal())
```
Add braces

shiltian commented on llvm/test/CodeGen/AMDGPU/swdev504645-global-fold.ll:
```diff
@@ -0,0 +1,46 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 5
+; RUN: llc -mtriple=amdgcn -mcpu=gfx942 < %s | FileCheck %s
```
```suggestion
; RUN: llc -mtriple=amdgcn-amd-amdhsa -mcpu=gfx942 < %s | FileCheck %s
```

isakhilesh replied:
Done

shiltian commented on llvm/test/CodeGen/AMDGPU/swdev504645-global-fold.ll:
```diff
@@ -0,0 +1,46 @@
+
+attributes #0 = { nocallback nofree nosync nounwind speculatable willreturn memory(none) }
+
+!0 = !{!"DescriptorBuffer", i32 4, i32 8, i32 0, i32 0}
```
do we really need these attribute?

isakhilesh replied:
I have removed the attribute code in line 44 but have kept the code in line 46 because without it there was an error asking for !0.
This file was reduced in llvm-reduce from an input .ll file generated by a fuzzer based on the ticket mentioned in the description.

isakhilesh committed:
Updated Patch, based on feedback.

dhruvachak reviewed (approved):
LGTM. Please wait for approval from @arsenm .

arsenm commented on llvm/test/CodeGen/AMDGPU/swdev504645-global-fold.ll:
```diff
@@ -0,0 +1,44 @@
+
+declare i32 @llvm.amdgcn.reloc.constant(metadata) #0
+
+!0 = !{!"DescriptorBuffer", i32 4, i32 8, i32 0, i32 0}
```
```suggestion
!0 = !{!"DescriptorBuffer", i32 4, i32 8, i32 0, i32 0}

```

arsenm committed:
fix whitespace error
